### DISCUSSION TITLE
RTH Fail. Crash.

### DISCUSSION THREAD

**Discussion by ekap12** [iNavFlight/inav#7826] [Q&A] [answered]
Almost lost my quadcopter today.
I checked the RTH mode, everything works (2:43) and already at (4:10) my battery in the Failsafe RTH quadcopter remote control runs down and it blows downwind to the first obstacle. What is my settings error?
  The failsafe settings are set to RTH. 
INAV 4.0.0

Video: https://www.youtube.com/watch?v=mjNhvFhblHQ
Blackbox: https://disk.yandex.ru/d/TkXJ_1dIIcUh2w
Diff Config: https://disk.yandex.ru/d/PJy0lyJxnoqDVg

**Comment by stronnag** [comment]
As discussed on telegram ... always suspicious when there is a mechanical fault; here we have two. Something causes a loss of navigation (sensor) integrity as 125.7s, then later we have the failsafe. What is also noteworthy at the end is that we have an RTH nav state, but not an RTH flight mode (239.4s), whereas during the commanded RTH (152.1s), RTH is asserted both as a Nav mode and flight mode. 
```
Iteration	Time(s)	Elapsed(s)	State	FltMode	navFlag
        0	  41.8	(   0.0)	nav_state_undefined (0)	0		
     5600	  44.9	(   3.1)	nav_state_althold_in_progress (3)	ARM|ANGLE|NAVALTHOLD		
    17888	  51.6	(   9.8)	nav_state_poshold_3d_in_progress (7)	ARM|ANGLE|NAVPOSHOLD		
   228096	 167.5	( 125.7)	nav_state_undefined (0)	ARM|ANGLE|NAVPOSHOLD		!AltTrust !PosTrust !HdgTrust
   228288	 167.6	( 125.8)	nav_state_poshold_3d_in_progress (7)	ARM|ANGLE|NAVPOSHOLD		
   276032	 194.0	( 152.1)	nav_state_rth_head_home (10)	ARM|ANGLE|NAVRTH|NAVPOSHOLD		
   305820	 210.4	( 168.6)	nav_state_rth_hover_prior_to_landing (11)	ARM|ANGLE|NAVRTH|NAVPOSHOLD		
   309460	 212.4	( 170.6)	nav_state_rth_landing (12)	ARM|ANGLE|NAVRTH|NAVPOSHOLD		
   316768	 216.4	( 174.6)	nav_state_poshold_3d_in_progress (7)	ARM|ANGLE|NAVPOSHOLD		
   434230	 281.2	( 239.4)	FAILSAFE=>RETURN_TO_HOME (1658,1908,1535,1580);(0,1)
   434230	 281.2	( 239.4)	nav_state_rth_head_home (10)	ARM|ANGLE|NAVPOSHOLD	
```
@breadoven, appreciate your thoughts  if you have time / interest. In particular that for the commanded RTH (152.1s) NAVRTH is asserted as a flight mode, and for the F/S instance (239.4s), it is not. The machine flies some 300m beyond home into the trees (so I don't think it's the 4.0 F/S emergency land issue this time).

![fsrth-crash](https://user-images.githubusercontent.com/158229/153729972-33f6b1ed-0ec0-4e64-bcb9-ec68e5e890c5.png)

**Reply by breadoven** [reply]
Doesn't flight mode just show which modes are selected by switched command ? RTH wasn't selected by switch during the FS so doesn't appear as a mode.

One obvious thing is the FS RTH navFlags show it's constantly adjusting position I assume because there is constant rcData input on pitch and roll (Rx FS appears to be set on hold ?). Wouldn't this override the Nav directional control back to the home point ? rcCommand values on RPY all end up stuck on 0 once the FS starts so it just drifted with no directional control. Would need to check the code to understand what's going on fully.

**Comment by breadoven** [comment, marked as answer]
Pretty sure this is caused by a bug here:

https://github.com/iNavFlight/inav/blob/c4bcd2ea2adde17915effcd32419e7a721a51f1c/src/main/navigation/navigation.c#L3137

If it's already adjusting position when the failsafe occurs `posControl.flags.isAdjustingPosition` remains true which overrides the RTH direction control. Needs resetting to false whenever FS is active although I'm not sure how that fits in with `resetMulticopterBrakingMode` @DzikuVx

**Reply by breadoven** [reply]
Fix provided by https://github.com/iNavFlight/inav/pull/7829.

Should this be raised as a new issue for traceability rather than a discussion item ?
